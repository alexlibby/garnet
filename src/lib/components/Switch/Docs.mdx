import { Canvas, Meta } from "@storybook/blocks";
import { ArgTypes } from "@storybook/blocks";
import * as SwitchStories from "./Switch.stories";

<Meta of={SwitchStories} />

# Switch

Switch is the primary component. It has 2 possible states.

- [Default](#default)
- [Disabled](#disabled)

## Default

This is the default state for our Switch component:

<Canvas of={SwitchStories.Default} />

## Disabled

This is the disabled state for our Switch component:

<Canvas of={SwitchStories.Disabled} />

## Properties of component

Below is a list of arguments available for this component:

<ArgTypes of={SwitchStories} />
