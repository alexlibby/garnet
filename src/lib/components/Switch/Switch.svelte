<svelte:options customElement="garnet-switch" />

<script>
  import { createEventDispatcher } from 'svelte';
  
  export let id = "";
  export let label = "";
  export let css = "";
  export let size = "";
  export let isChecked = false;
  export let disabled = false;

  const dispatch = createEventDispatcher();

  function handleChange(event) {
    const { checked } = event.target;
    dispatch('change', { checked });
  }

  $: switchContainer = [
      "switch-container",
      css ? css : "",
      disabled ? "disabled" : "",
    ].filter(c => c).join(" ");
  
  let classes = ["switch", size ? `switch-${size}` : ""].join(" ");
</script>

<label class={switchContainer} for={id}>
  <input
    type="checkbox"
    class="switch-input"
    id={id}
    bind:checked={isChecked}
    disabled={disabled}
    on:change={handleChange}
    role="switch"
  />
  <span class={classes}></span>
  <span class="switch-label">{label}</span>
</label>

<style>
  .switch-container {
    display: flex;
    min-height: 36px;
    width: 100%;
    padding: 8px;
    position: relative;
    align-items: center;
  }
  
  .switch-container:hover {
    cursor: pointer;
  }
  
  .switch::before,
  .switch::after {
    border: 1px solid #ccc;
    content: "";
    position: absolute;
    top: 50%;
    transform: translateY(-50%);
  }
  
  .switch::after {
    background: #fff;
    border-radius: 100%;
    width: 22.4px;
    height: 22.4px;
    right: 22.4px;
    transition: right 200ms ease-in-out;
  }
  
  .switch::before {
    background: #eee;
    border-radius: 28px;
    width: 44px;
    height: 28px;
    right: 4px;
    transition: background 300ms ease-in-out;
  }
      
  .switch-input:checked + .switch::after {
    right: 6.4px;
  }

  .switch-input {
    margin: 0;
    opacity: 0.01%;
    position: absolute;
    left: 0;
    top: 0;
    width: 100%;
    height: 100%;
    pointer-events: none;
  }
  
  .switch-input:focus + .switch::before {
    box-shadow: 0 0 0 3px rgb(55 149 225 / 50%);
  }
  
  .switch-input:checked + .switch:not(.switch-border)::before {
    background: #733635;
  }

  .switch-input[disabled] + .switch,
  .switch-input[disabled] + .switch-label,
  .switch-container.disabled {
    color:#717171;
    appearance: none;
    box-shadow: none;
    cursor: not-allowed;
    opacity: 80%;
  }
  </style>