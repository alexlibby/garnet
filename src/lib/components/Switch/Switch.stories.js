import Switch from "./Switch.svelte";
import { fn } from "@storybook/test";
import { action } from "@storybook/addon-actions";
import SwitchDecorator from "./SwitchDecorator.svelte";
import { BADGE } from "storybook-addon-badges";

const BADGES = {
  ...BADGE,
  ALPHA: "Alpha",
};

export default {
  title: "Garnet UI Library/Animation Components/Switch",
  component: Switch,
  decorators: [() => SwitchDecorator],
  argTypes: {
    disabled: { control: "boolean" },
    label: { control: "string" },
  },
  on: { change: fn().mockName("on-change") },
};

export const Default = (args) => ({
  Component: Switch,
  props: {
    ...args,
    id: "1",
    label: "Switch with label",
    disabled: false,
  },
  on: { change: fn().mockName("on-change") },
});

Default.parameters = {
  badges: [BADGES.ALPHA],
  docs: { source: { type: "code" } },
};

export const Disabled = (args) => ({
  Component: Switch,
  props: {
    ...args,
    id: "1",
    label: "Disabled switch with label",
    disabled: true,
  },
  on: { change: fn().mockName("on-change") },
});

Disabled.parameters = {
  badges: [BADGES.ALPHA],
  docs: { source: { type: "code" } },
};
