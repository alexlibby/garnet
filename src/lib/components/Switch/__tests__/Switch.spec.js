import { describe, test, expect } from "vitest";
import { axe } from "vitest-axe";
import * as AxeMatchers from "vitest-axe/matchers";
import { render, screen, fireEvent } from "@testing-library/svelte";
import Switch from "../../Switch/Switch.svelte";

const SwitchHTML = `<div class="garnet s-4Y75ZjC11XWf"><dialog class="alert alert-info fade-in s-4Y75ZjC11XWf" role="alertdialog"  title="Alert Dialog"><div class="icon s-4Y75ZjC11XWf"><svg width="24px" height="24px" viewBox="0 0 512 512"><path d="M248 64C146.39 64 64 146.39 64 248s82.39 184 184 184 184-82.39 184-184S349.61 64 248 64z" fill="none" stroke="currentColor" stroke-miterlimit="10" stroke-width="32"></path><path fill="none" stroke="#ffffff" stroke-linecap="round" stroke-linejoin="round" stroke-width="32" d="M220 220h32v116"></path><path fill="none" stroke="currentColor" stroke-linecap="round" stroke-miterlimit="10" stroke-width="32" d="M208 340h88"></path><path d="M248 130a26 26 0 1026 26 26 26 0 00-26-26z" fill="#ffffff"></path></svg><!--<Icon>--></div> <div class="message s-4Y75ZjC11XWf"><strong class="s-4Y75ZjC11XWf">Simple Info</strong> An info description</div> <div class="s-4Y75ZjC11XWf"><button class="s-4Y75ZjC11XWf">âœ–</button></div></dialog></div>`;
expect.extend(AxeMatchers);

describe("Tests for Switch", () => {
  const props = {
    disabled: false,
    label: "Disabled switch with label",
  };

  test("should render properly", () => {
    const result = render(Switch, { props });
    expect(() => result).not.toThrow();
  });

  test("show a label with correct text", () => {
    const result = render(Switch, { props });
    expect(result).not.toBeNull();
  });

  test("should show a switch with enabled background when clicked", async () => {
    render(Switch, { props });
    const SwitchExample = screen.getByRole("switch");
    await fireEvent.click(SwitchExample);
    expect(SwitchExample.checked).toBe(true);
  });

  test("get a snapshot of component", () => {
    // Create a new container for the test and mount component
    const host = document.createElement("div");
    document.body.appendChild(host);
    const instance = new Switch({ target: host, props: props });

    // check that all rendered OK
    expect(instance).toBeTruthy();

    // Take snapshot
    expect(host.innerHTML).toMatchSnapshot();
  });

  test("should demonstrate no issues with accessibility", async () => {
    const render = () => SwitchHTML;
    const result = render();

    expect(await axe(result)).toHaveNoViolations();
  });
});
