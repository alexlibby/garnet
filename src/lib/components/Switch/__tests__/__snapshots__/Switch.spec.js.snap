// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Tests for Switch > get a snapshot of component 1`] = `"<label class="switch-container svelte-1wtopcf" for=""><input type="checkbox" class="switch-input svelte-1wtopcf" id="" role="switch"> <span class="switch  svelte-1wtopcf"></span> <span class="switch-label">Disabled switch with label</span></label>"`;
